presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -975,12 +974,9 @@ else if (criteria instanceof JoinOn) {
 
                 Analyzer.verifyNoAggregatesOrWindowFunctions(metadata.getFunctionRegistry(), expression, "JOIN clause");
 
-                Expression canonicalized = CanonicalizeExpressions.canonicalizeExpression(expression);
-                analyzeExpression(canonicalized, output);
-
                 Set<Expression> postJoinConjuncts = new HashSet<>();
 
-                for (Expression conjunct : ExpressionUtils.extractConjuncts(canonicalized)) {
+                for (Expression conjunct : ExpressionUtils.extractConjuncts(expression)) {
                     conjunct = ExpressionUtils.normalize(conjunct);
```

Comment:
I think this is also problematic. Just like `CanonicalizeExpressions.canonicalizeExpression(expression);` rewrites expression tree, `ExpressionUtils.normalize(conjunct);` also rewrites the tree. Thus resulting `conjuct` might not have coercions (and types) registered in analysis. I think it's enough to reanalyze `conjuct` expression. It already happens (partially) at line: 1007.

Anyway, I am thinking that the whole section:
```
 for (Expression conjunct : ExpressionUtils.extractConjuncts(expression)) {
   ...
 }
```
might not be required.

It seems that the only thing it actually does at the end is: `analysis.recordSubqueries(node, analyzeExpression(postJoinPredicate, output));`

So maybe the whole section can be rewritten as:
```
analysis.recordSubqueries(node, expressionAnalysis);
```
and it should work :)

I would give it a try and check if tests don't fail. Getting rid of the complex code below would be a big plus.